react/features/stream-effects/blur/JitsiStreamCropPersonEffect.js

```diff
@@ -0,0 +1,193 @@
+                : desktopTrack.getConstraints();
+        this._height = height;
+        this._width = width;
+        this._frameRate = frameRate;
+        this._inputDesktopElement.width = width;
+        this._inputDesktopElement.height = height;
+        this._inputDesktopElement.autoplay = true;
+        this._inputDesktopElement.srcObject = desktopStream;
+
+        const videoDiv = document.createElement("div");
+        document.body.appendChild(videoDiv);
+        videoDiv.appendChild(this._inputVideoElement);
+        videoDiv.appendChild(this._inputDesktopElement);
+        // set the style attribute of the div to make it invisible
+        videoDiv.style.display = "none";
+
+        this._maskCanvasElement.width = this._cameraWidth;
+        this._maskCanvasElement.height = this._cameraHeight;
+        this._maskCanvasContext = this._maskCanvasElement.getContext('2d');
+        this._outputCanvasElement.width = this._width;
+        this._outputCanvasElement.height = this._height;
+
+        this._videoFrameTimerWorker.postMessage({
+            id: SET_INTERVAL,
+            timeMs: 1000 / this._cameraFrameRate
+        });
+        this._maskFrameTimerWorker.postMessage({
+            id: SET_INTERVAL,
+            timeMs: 50
+        });
+
+        return this._outputCanvasElement.captureStream(this._cameraFrameRate);
+    }
+
+    /**
+     * Stops the capture and render loop.
+     *
+     * @returns {void}
+     */
+    stopEffect() {
+        this._videoFrameTimerWorker.postMessage({
+            id: CLEAR_INTERVAL
+        });
+        this._maskFrameTimerWorker.postMessage({
+            id: CLEAR_INTERVAL
+        });
+    }
+
+    /**
+     * Loop function to render the video frame input and draw blur effect.
+     *
+     * @private
+     * @returns {void}
+     */
+    _renderVideo() {
+        this._maskCanvasContext.drawImage(this._inputVideoElement, 0, 0,
+            this._cameraWidth, this._cameraHeight);
+        if (this._segmentationData) {
+            const img = this._maskCanvasContext.getImageData(0, 0, 
+                this._cameraWidth, this._cameraHeight);
+            let pixels = img.data;
+            const data = this._segmentationData.data;
+            for (let i = 0; i < data.length; i++) {
+                // if the pixel doesn't belong to the person, set the 
+                // alpha channel value to 0 to make it transparent
+                pixels[i * 4 + 3] = data[i] == 0 ? 0 : 255;
+            }
+            this._maskCanvasContext.putImageData(img, 0, 0);
+        }
+        this._ctx.drawImage(this._inputDesktopElement, 0, 0, this._width,
+            this._height);
+        this._ctx.drawImage(this._maskCanvasElement, this._width - this._cameraWidth,
+            this._height - this._cameraHeight, this._cameraWidth, this._cameraHeight);
+    }
+
+    /**
+     * Loop function to render the background mask.
+     *
+     * @private
+     * @returns {void}
+     */
+    _renderMask() {
```

Comment:
How does this method do the actual rendering? Is something listening for changes to this._segmenetationData?

Reply:
_renderVideo() is called every 1000/frameRate ms, the mask is calculated using the available segmentation data and applied on top of the desktop track. The segmentation data is calculated every 50ms.